File: trees/balanced_binary_tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Boolean}
def is_balanced(root)
  return true if root.nil?

  (get_height(root.left) - get_height(root.right)).abs <= 1 &&
  is_balanced(root.left) &&
  is_balanced(root.right)
end

def get_height(root)
  return 0 if root.nil?
  1 + [get_height(root.left), get_height(root.right)].max
end

# Faster solution: Keep checking for condition as we're calculating height
def is_balanced(root)
  dfs_height(root) != -1
end

def dfs_height(root)
  return 0 if root.nil?
  l = dfs_height(root.left)
  r = dfs_height(root.right)
  return -1 if l == -1          # left is balanced
  return -1 if r == -1          # right is balanced
  return -1 if (l - r).abs > 1  # height difference for left and right is at most 1

  1 + [l, r].max
end



# https://leetcode.com/problems/balanced-binary-tree/description/
# 110. Balanced Binary Tree
# Test on web site directly

File: trees/balanced_binary_tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def isBalanced(self, root: TreeNode) -> bool:
        return self.dfs(root)[0]

    def dfs(self, node):
        if not node: return True, 0

        l_bool, l_depth = self.dfs(node.left)
        if not l_bool: return False, None

        r_bool, r_depth = self.dfs(node.right)
        if not r_bool: return False, None

        return abs(l_depth - r_depth) <= 1, max(l_depth, r_depth) + 1
